src/physics/utilities/boundary_condition.cpp

```diff
@@ -48,18 +48,19 @@ void BoundaryCondition::project(FiniteElementState& state) const
     std::transform(tdofs.begin(), tdofs.end(), dof_list.begin(),
                    [&space = std::as_const(state.space())](int tdof) { return space.VDofToDof(tdof); });
 
-    if (component_ == -1) {
-      // If it contains all components, project the vector
-      auto vec_coef = std::get_if<std::shared_ptr<mfem::VectorCoefficient>>(&coef_);
-      SLIC_ASSERT_MSG(vec_coef,
-                      "Essential boundary condition contained all components but had a non-vector coefficient.");
-      state.gridFunc().ProjectCoefficient(**vec_coef, dof_list);  // Deref ptr-to-result, then shared_ptr
+    // the only reason to store a VectorCoefficient is to act on all components
+    if (std::holds_alternative<std::shared_ptr<mfem::VectorCoefficient>>(coef_)) {
+      auto vec_coef = std::get<std::shared_ptr<mfem::VectorCoefficient>>(coef_);
+      state.gridFunc().ProjectCoefficient(*vec_coef, dof_list);
     } else {
-      // If it is only a single component, project the scalar
-      auto scalar_coef = std::get_if<std::shared_ptr<mfem::Coefficient>>(&coef_);
-      SLIC_ASSERT_MSG(scalar_coef,
-                      "Essential boundary condition contained a single component but had a non-scalar coefficient.");
-      state.gridFunc().ProjectCoefficient(**scalar_coef, dof_list, component_);
+      // an mfem::Coefficient could be used to describe a scalar-valued function, or
+      // a single component of a vector-valued function
+      auto scalar_coef = std::get<std::shared_ptr<mfem::Coefficient>>(coef_);
+      if (component_ == -1) {
+        state.gridFunc().ProjectCoefficient(*scalar_coef, dof_list);
+      } else {
+        state.gridFunc().ProjectCoefficient(*scalar_coef, dof_list, component_);
+      }
```

Comment:
I ran into this in #287 and solved it [slightly differently](https://github.com/LLNL/serac/pull/287/files#diff-dd9bd49c73ab278f257c5c3c0c11dcf9e60388f04d4484a64bc0ade8a8d55387R57-R70), but I think I like this approach more.  Maybe they could be combined by condensing 59-63 with a ternary operator?

Reply:
changed to
```
state.gridFunc().ProjectCoefficient(*scalar_coef, dof_list, std::max(0, component_));
```